apps/server/src/form-instances/form-instances.service.ts

```diff
@@ -99,4 +105,45 @@ export class FormInstancesService {
       },
     });
   }
+
+  async signFormInstance(formInstanceId: string, signatureId: string) {
+    const formInstance = await this.prisma.formInstance.findUnique({
+      where: { id: formInstanceId },
+      include: {
+        signatures: { include: { signerPosition: true, userSignedBy: true } },
+      },
+    });
+
+    if (!formInstance) {
+      throw new NotFoundException(
+        FormInstanceErrorMessage.FORM_INSTANCE_NOT_FOUND,
+      );
+    }
+
+    const signatureIndex = formInstance.signatures.findIndex(
+      (sig) => sig.id === signatureId,
+    );
+
+    if (signatureIndex === -1) {
+      throw new NotFoundException(SignatureErrorMessage.SIGNATURE_NOT_FOUND);
+    }
+
+    for (let i = 0; i < signatureIndex; i++) {
+      if (!formInstance.signatures[i].signed) {
+        throw new BadRequestException(SignatureErrorMessage.SIGNATURE_NOT_NEXT);
+      }
+    }
+
+    const updatedSignature = await this.prisma.signature.update({
+      where: { id: signatureId },
+      data: { signed: true },
+    });
+
+    formInstance.signatures[signatureIndex] = {
+      ...formInstance.signatures[signatureIndex],
```

Comment:
is this aggregating the existing signature with the new updates?

Reply:
Yes, exactly. It's meant to update the signatures at the signatureIndex accordingly.